### ISSUE TITLE
Operator connection fails with 431 Request Header Fields Too Large

### ISSUE THREAD

**Issue by dbarentine** [metalbear-co/mirrord#2172]
### Bug Description

When I try to run mirrord using via the operator I get the following error:
Failed to connect to the operator, probably due to RBAC: creating a websocket connection failed: failed to upgrade to a WebSocket connection: failed to switch protocol: 431 Request Header Fields Too Large

![image](https://github.com/metalbear-co/mirrord/assets/939865/167f1630-0aea-42ee-b903-c0be6277439c)

### Steps to Reproduce

1. Installed Mirrord operator
2. Installed Jetbrains Rider Mirrord plugin
3. Tried to run Mirrord using the operator (as we want to use teams) - I tried this both in Rider and via the CLI. Both failed with the same error.

Config file looks like:
```json
{
    "target": {
        "path": {
            "deployment": "auditsupervisor-dev"
        },
        "namespace": "platform-dev"
    },
    "agent": {
        "namespace": "platform-dev"
    },
    "operator": true,
    "feature": {
        "network": {
            "incoming": {
                "mode": "steal",
                "port_mapping": [[ 5000, 8080 ]],
                "http_filter": {
                    "header_filter": "X-Starling-Route-Key: dane"
                }
            },
            "outgoing": {
                "tcp": false,
                "udp": false
            },
            "dns": false
        },
        "fs": {
            "mode": "local"
        },
        "env": {
            "include": ""
        }
    }
}
```

Operator Helm Chart Version: mirrord-operator-0.10.1
Operator Version: 3.72.0
Kubernetes Version: 1.28.3
Kubectl Version: 1.29.0
Rider Plugin Version 3.57.0

We do have Istio 1.20.0 running in the cluster. I disabled the sidecar on the mirrord-operator pod to test but I still end up with the same error. I can't disable the sidecar on the actual pod I'm trying to steal from for other reasons.

If I don't use the operator I can run mirrord just fine against this same workload.

### Your operating system and version

Ubuntu 22.04 and Mirrord operator 3.72.0

### Local process

.NET 6 web application

**Comment by aviramha** [comment]
hey @dbarentine - sorry to hear about your experience. Hopefully we'll get it resolved together!
Few things to help us debug it better:
1. In the mirrord-operator deployment add env with key `RUST_LOG` set to `mirrord=trace` - please re-produce and attach logs.
2. Run `mirrord operator status` from cli - does that work?
3. Do you have any special configuration in your cluster? is it EKS/AKS/GKE/bare-metal/k3s/... etc? What version of k8s is used?
4. Do you have `HTTPS_PROXY` or `HTTP_PROXY` or `SOCKS_PROXY` in use?

Lastly, if you'd like we can schedule a debug session so we'll reduce the ping-pong time :) - https://calendly.com/aviram-5/30min

**Comment by dbarentine** [comment]
@aviramha I don't get any additional log messages with mirrord=trace. I changed it so the operator also is set to trace and think I may have an idea of what is going on. I've only included the mirrord=trace log below because the other one has some information I don't want to post here. But if you want me to schedule a debug session I can get you the full one.

[trace.txt](https://github.com/metalbear-co/mirrord/files/13942773/trace.txt)

```mirrord operator status```
![image](https://github.com/metalbear-co/mirrord/assets/939865/b37404bf-a230-4736-bcdc-eb7649e67ce6)

```Kubernetes Version```
![image](https://github.com/metalbear-co/mirrord/assets/939865/10b48fdd-5ea7-445c-8f98-938e36f5eb13)

No proxy

But I think this is the key, we are running AKS with Azure AD integration. I see this message from the operator trace:

[2;3min[0m operator_controller::restful::[1mhandle_get[0m [2;3mwith[0m [1mremote[0m: RestfulRemote { session_id: 187B6E73C72D6641 }, [1mpath[0m: {"namespace": "platform-dev", "name": "deploy.auditsupervisor-dev"}, [1mparams[0m: RestfulGetParams { connect: None, on_concurrent_steal: None }, [1mreq[0m: Request { method: GET, uri: https://10.240.13.95:3000/deploy.auditsupervisor-dev, version: HTTP/2.0, **headers: {"x-remote-group": "b1357603-...", "x-remote-group": "caeeaf03-...", "x-remote-group" ...**

I have a bunch of groups in AAD that appear to be getting included as an x-remote-group header. So the key to reproducing it is probably an AKS cluster that is AAD integrated with a user in a bunch of AAD groups.

Let me know if that's enough or if you still want me to schedule a debugging session.

**Comment by aviramha** [comment]
hi @dbarentine - I think it's a good clue - but I'm wondering what's the solution here - I'd suggest to increase the header size of kubeapi, as I assume those headers are needed for authentication? I'd be happy to schedule a session to debug it further and find a proper solution for this.

**Comment by dbarentine** [comment]
I think the problem with changing something at the cluster level is this is a managed cluster. So there is a good possibility that I have no ability to do so as Microsoft manages the k8s API control plane. But if you know how to change that setting let me know. I didn't see anything in the kube-apiserver documentation where that header comes up.

I'm not sure it's at that level anyways. We use this all the time and have never seen this particular issue across any of the third-party operators we've installed.

My suspicion is there is a limit set somewhere in one of the libraries being leveraged by the mirrord operator code. But I'm not all that familiar with Rust so it's hard for me to track down where it might possibly be. It feels like maybe it's in kube or hyper. Although hyper says it supports header maps up to 32 KB and I doubt I'm over that. Maybe tokio-tungstenite since the error specifically mentions it got the 431 on trying to upgrade to websockets. Although even that looks like a max of 64 KB. Given the output in the log I seem to be right around 8 KB.

I'm happy to schedule a debug session. But my trial just expired so let me get that straightened out first and then I will schedule it.

While I straighten that out maybe you could tell me where I find the operator/controller/src/restful.rs file? It's not under the operator folder in the mirrord repo. That's one of the files referenced when the operator logs the request with all the x-remote-group headers. I was hoping to take a look at it to try and understand how it works.
![image](https://github.com/metalbear-co/mirrord/assets/939865/63b77ffa-0b96-41dc-89da-3994115270b9)

**Comment by aviramha** [comment]
operator/controller/.. is part of a private repository.
I think you're right - this might be it
https://github.com/snapview/tungstenite-rs/pull/379/files

**Comment by aviramha** [comment]
Turns out hyper has a max header count, which is set to 100.
We've changed it to 1000 until they merge the configurable value PR
https://github.com/metalbear-co/hyper/commit/b61858cf6cf9ae6580cb9d14fefbed7291daf15c
It will be released soon as operator 3.73.1

**Comment by dbarentine** [comment]
Thanks @aviramha. I've upgraded the operator and it all seems to be working now. Thank you for the help and the quick fix.
